From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Fri, 30 Oct 2026 01:31:48 -0300
Subject: [PATCH] Fix MC-117075: TE Unload Lag Spike

We replaced the `blockEntityTickers` list with a custom list based on fastutil's `ObjectArrayList` with a small yet huge change for us: A method that allows us to remove a list of indexes from the list.

This is WAY FASTER than using `removeAll` with a list of entries to be removed, because we don't need to calculate the identity of each block entity to be removed, and we can jump directly to where the search should begin, giving a performance boost for small removals (because we don't need to loop thru the entire list to find what element should be removed) and a performance boost for big removals (no need to calculate the identity of each block entity).

diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index d49c024c024ba4ce8f0287b4bbccc80e1b319805..11156e3f9a095c077215d9f2c5a7b1b72bd57ed2 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -117,7 +117,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         .add(new ExplosionParticleInfo(ParticleTypes.POOF, 0.5F, 1.0F))
         .add(new ExplosionParticleInfo(ParticleTypes.SMOKE, 1.0F, 1.0F))
         .build();
-    public final List<TickingBlockEntity> blockEntityTickers = Lists.newArrayList();
+    public final net.sparklypower.sparklypaper.BlockEntityTickersList blockEntityTickers = new net.sparklypower.sparklypaper.BlockEntityTickersList(); // SparklyPaper - optimize block entity removals
     protected final CollectingNeighborUpdater neighborUpdater;
     private final List<TickingBlockEntity> pendingBlockEntityTickers = Lists.newArrayList();
     private boolean tickingBlockEntities;
@@ -1498,14 +1498,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         boolean tickBlockEntities = this.tickRateManager().runsNormally();
 
         int tickedEntities = 0; // Paper - rewrite chunk system
-        // Paper start - Fix MC-117075 use removeAll
-        final it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<@Nullable TickingBlockEntity> toRemove = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>();
-        toRemove.add(null);
-        for (int tickerIndex = 0; tickerIndex < this.blockEntityTickers.size(); tickerIndex++) {
+        for (int tickerIndex = 0; tickerIndex < this.blockEntityTickers.size(); tickerIndex++) { // Paper - Fix MC-117075 use removeAll
             final TickingBlockEntity ticker = this.blockEntityTickers.get(tickerIndex);
-        // Paper end - Fix MC-117075 use removeAll
             if (ticker.isRemoved()) {
-                toRemove.add(ticker); // Paper - Fix MC-117075 use removeAll
+                this.blockEntityTickers.markAsRemoved(tickerIndex); // SparklyPaper - optimize block entity removals
             } else if (tickBlockEntities && this.shouldTickBlocksAt(ticker.getPos())) {
                 ticker.tick();
                 // Paper start - rewrite chunk system
@@ -1516,7 +1512,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             }
         }
 
-        this.blockEntityTickers.removeAll(toRemove); // Paper - Fix MC-117075 use removeAll
+        this.blockEntityTickers.removeMarkedEntries(); // SparklyPaper - optimize block entity removals
         this.tickingBlockEntities = false;
     }
 
